ashpd-demo/src/portals/desktop/usb.rs

```diff
@@ -0,0 +1,373 @@
+
+glib::wrapper! {
+    pub struct UsbDeviceRow(ObjectSubclass<imp::UsbDeviceRow>)
+        @extends gtk::Widget, gtk::ListBoxRow, adw::PreferencesRow, adw::ActionRow;
+}
+
+impl UsbDeviceRow {
+    fn with_device(page: UsbPage, device_id: String, writable: bool) -> Self {
+        let obj: Self = glib::Object::new();
+
+        let imp = obj.imp();
+        imp.page.replace(Some(page));
+        imp.device_id.replace(device_id);
+        imp.writable.set(writable);
+        obj
+    }
+
+    fn acquire(&self) {
+        self.imp().checkbox.set_active(true);
+    }
+
+    fn release(&self) {
+        self.imp().checkbox.set_active(false);
+    }
+}
+
+mod imp {
+    use std::cell::{Cell, RefCell};
+
+    use super::*;
+
+    #[derive(Debug, gtk::CompositeTemplate, Default)]
+    #[template(resource = "/com/belmoussaoui/ashpd/demo/usb_device_row.ui")]
+    pub struct UsbDeviceRow {
+        #[template_child]
+        pub(super) checkbox: TemplateChild<gtk::CheckButton>,
+        #[template_child]
+        pub(super) acquire: TemplateChild<gtk::Button>,
+        #[template_child]
+        pub(super) release: TemplateChild<gtk::Button>,
+        pub(super) page: RefCell<Option<super::UsbPage>>,
+        pub(super) device_id: RefCell<String>,
+        pub(super) writable: Cell<bool>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for UsbDeviceRow {
+        const NAME: &'static str = "UsbDeviceRow";
+        type Type = super::UsbDeviceRow;
+        type ParentType = adw::ActionRow;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.bind_template();
+            klass.bind_template_callbacks();
+        }
+
+        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    #[gtk::template_callbacks]
+    impl UsbDeviceRow {
+        #[template_callback]
+        async fn handle_acquire_clicked(&self, _: &gtk::Button) {
+            let page = {
+                self.page.borrow().clone()
+            };
+            if let Some(page) = page {
+                let device_id = self.device_id.borrow().clone();
+                let writable = self.writable.get();
+                page.share(&device_id, writable).await;
+            }
+        }
+
+        #[template_callback]
+        async fn handle_release_clicked(&self, _: &gtk::Button) {
+            let page = {
+                self.page.borrow().clone()
+            };
+            if let Some(page) = page {
+                let device_id = self.device_id.borrow().clone();
+                page.unshare(&device_id).await;
+            }
+        }
+    }
+
+    impl ObjectImpl for UsbDeviceRow {}
+    impl WidgetImpl for UsbDeviceRow {}
+    impl ListBoxRowImpl for UsbDeviceRow {}
+    impl PreferencesRowImpl for UsbDeviceRow {}
+    impl ActionRowImpl for UsbDeviceRow {}
+
+    #[derive(Debug, gtk::CompositeTemplate, Default)]
+    #[template(resource = "/com/belmoussaoui/ashpd/demo/usb.ui")]
```

Comment:
and split different types in a separate mod, see https://github.com/bilelmoussaoui/ashpd/blob/master/ashpd-demo/src/portals/desktop/notification.rs#L161